test/ml_dsa.inc

```diff
@@ -0,0 +1,1800 @@
+/*
+ * Copyright 2024 The OpenSSL Project Authors. All Rights Reserved.
+ *
+ * Licensed under the Apache License 2.0 (the "License").  You may not use
+ * this file except in compliance with the License.  You can obtain a copy
+ * in the file LICENSE in the source distribution or at
+ * https://www.openssl.org/source/license.html
+ */
+
+#define ML_DSA_SIG_GEN_TEST_DET_ITEM(alg, name) {                              \
+    alg,                                                                       \
+    name##_priv, sizeof(name##_priv),                                          \
+    name##_msg, sizeof(name##_msg),                                            \
+    name##_digest, sizeof(name##_digest),                                      \
+}
+
+#define ML_DSA_SIG_GEN_TEST_ITEM(alg, name) {                                  \
+    alg,                                                                       \
+    name##_priv, sizeof(name##_priv),                                          \
+    name##_msg, sizeof(name##_msg),                                            \
+    name##_digest, sizeof(name##_digest),                                      \
+    name##_add_random, sizeof(name##_add_random),                              \
+}
+
+#define ML_DSA_KEYGEN_TEST_ITEM(alg, name) {                                   \
+    alg,                                                                       \
+    name##_keygen_seed, sizeof(name##_keygen_seed),                            \
+    name##_keygen_pub, sizeof(name##_keygen_pub),                              \
+    name##_keygen_priv, sizeof(name##_keygen_priv),                            \
+}
+
+typedef struct ml_dsa_sig_test_data_st {
+    const char *alg;
+    const unsigned char *priv;
+    size_t priv_len;
+    const unsigned char *msg;
+    size_t msg_len;
+    /* This is the sha256 digest of the signature, to reduce its size */
+    const unsigned char *sig_digest;
+    size_t sig_digest_len;
+    const unsigned char *add_random;
+    size_t add_random_len;
+} ML_DSA_SIG_TEST_DATA;
+
+typedef struct ml_dsa_keygen_test_data_st {
+    const char *name;
+    const uint8_t *seed;
+    size_t seed_len;
+    const uint8_t *pub;
+    size_t pub_len;
+    const uint8_t *priv;
+    size_t priv_len;
+} ML_DSA_KEYGEN_TEST_DATA;
+
+static const uint8_t ml_dsa_65_0_sig_gen_priv[] = {
```

Comment:
The *origin* of the test data should be noted - where did these come from ...

Reply:
Either way they need to be autogenerated. So this still requires some work that needs to be done, but not as part of this PR.
I wrote some relatively simple python code yesterday that takes a json file as input and autogenerates the tables.
This could be adapted to do whatever we want it to - it just takes time.
I am not overly keen on the process of manually copying large data fields from a json file.

These are acvp tests so they dont really fit into evp_test very well.. But this is work for another day.